### PR TITLE
Record unicode cmap encodings when one glyph is in multiple slots

### PR DESCRIPTION
**Pull request by skef**
The problem addressed by this PR is referenced in multiple bugs, including #1534, #1831, #3085 and possibly #3080. However, the scope of the problem and the complete fix is as yet unclear.

### CODE REVIEW

**Comment by skef** [comment]
@HinTak

I know there is a complex and perhaps unpleasant history with this family of issues but I'm hoping you can help me out with taking a look at a potential fix. Keep in mind that I know *next to nothing* about CID mapped fonts so you might have to slowly walk me through steps like verifying problems and/or fixes in the UI. 

Since you've made at least one PR I assume you can compile FontForge. The code for this PR is in this branch: https://github.com/skef/fontforge/tree/bg1831 , which you can look at by cloning `https://github.com/skef/fontforge` and doing a `git checkout bg1831` "inside" the repository. 

Based on preliminary testing (of `Open`/`Generate` pairs) this seems to resolve the noted problem in #1831 and I suspect it gets most or all of the way with #1534. But again I'm very green with this CID stuff.

**Comment by skef** [comment]
@HinTak I should have supplied more context about this fix.

The "altuni" functionality and interface already exists. The main UI is in the "Alternate Unicode Encodings" section of the "Unicode" (first) tab of the "Glyph Info..." dialog. The python interface is described [here](https://fontforge.org/docs/scripting/python/fontforge.html#fontforge.glyph.altuni). (I don't know if there is a native FontForge scripting interface -- at first glance there doesn't seem to be.)

All this PR does is add the code to make use of it on ttf/otf import. I don't know why it wasn't doing so before except that looking in the repository history the code was very early and perhaps GW never got around to addressing the case.

I'm pretty sure there's already some "deduplication" code present, much like there is for   single-encoding cases. You'll have to play with it a bit. Regardless, most of what you've mentioned is already in the code base.

**Comment by skef** [comment]
There's more info about the UI [here](https://fontforge.org/docs/ui/dialogs/charinfo.html). I may not currently be setting all the variables correctly. (It looks like 0 would have been a better choice than -1 for Variation Selector? Or maybe not given that this is a CID font being loaded? It's a little hard to say because it's a bit of one and a bit of the other.).

(Also, to make this a complete response altuni data are already stored in the sfd file.)

**Comment by HinTak** [comment]
Okay... A few years ago when I looked into this I saw the altuni struct and couldn't figure out what it is supposed to do. Maybe it is just unfinished code.

**Comment by skef** [comment]
@HinTak These CID/altuni problems are unlikely to ever be fixed without test cases, or at least detailed guidance, from someone with more familiarity with the (practical) issues and problems. I've already run into at least one gap in the altuni implementation ( #4598 ) but I don't have a good picture for what more, if anything, is missing. 

You seem to be the main interested party so far. If we work together I think it's likely we could make some progress. Otherwise this is likely to sit in the PR queue and eventually get deleted when @ctrlcctrlv does a cleanup pass.

**Comment by HinTak** [comment]
It is relatively simple to test the correctness of the change: the Adobe source cjk fonts are known to be exhibit this kind of behavior, and you should be able to get at the multiple mappings via the new code and compare what my freetype-py script does. I just don't have either the time nor the inclination to build fontforge from source at the moment.

**Comment by frank-trampe** [comment]
@skef, this pull request seems to fix real problems, and I see no significant risk of regression from it. Is there any reason we couldn't merge it now?

**Comment by skef** [comment]
Yeah, I guess the likelihood of this causing problems is low. The risk is not tracking what else is needed but that's always a risk. 

It would be much better if we had a full idea of how the altuni system should work with test cases.

**Comment by HinTak** [comment]
It should be possible to rewrite part of, and compare with the other freetype-py dependent script I wrote without freetype-py .

**Comment by skef** [comment]
I looked at this a bit more. Fixing the altuni stuff more completely probably requires an overhaul of the Encoding code. At the very least it would require going through that stuff with a fine tooth comb. So by the "improves things and doesn't cause obvious harm" standard this could probably go in.

**Comment by HinTak** [comment]
As I replied earlier some months ago, it should be possible to re-write my freetype-py dependent script with the new functionality and remove the freetype-py dependency, and also run the two side by side to verify that the new functionality works correctly (if you provide a python binding to the new functionality).

**Comment by skef** [comment]
@HinTak The current issues have less to do with bottom-level glyph state than with having the GUI update correctly when manipulated via scripting.

File: fontforge/parsettf.c [hunk 1]
```diff
@@ -4558,6 +4558,30 @@ static int PickCMap(struct cmap_encs *cmap_encs,int enccnt,int def) {
...
 }
 
+/* This doesn't worry about putting the most appropriate encoding
+   in sc->unicodeenc because (in theory) NameConsistancyCheck()
+   will swap it in. */
+static void addttfencoding(SplineChar *sc, int unc) {
+    struct altuni *alt;
+    if ( unc==-1 ) {
+	return;
+    } else if (sc->unicodeenc==-1 || sc->unicodeenc==unc) {
```
Comment (jtanx) [inline fontforge/parsettf.c:4568]: Seems like you could move `sc->unicodeenc==unc` to the if statement above

**Review by jtanx — APPROVED** [review]

File: fontforge/parsettf.c (hunk 1 above)
Comment (skef) [reply fontforge/parsettf.c]: done

**Comment by jtanx** [comment]
Change looks fine but needs to be rebased to fix conflicts with master

**Comment by skef** [comment]
Rebased

**Pull request merged**